source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+            m_writer->emit(attr->getOffset());
+            m_writer->emit(") ");
+            m_writer->emit("@group(");
+            m_writer->emit(attr->getSpace());
+            m_writer->emit(") ");
+            break;
+
+        }
+
+    }
+
+    auto elementType = type->getElementType();
+    m_writer->emit("var<uniform> ");
+    m_writer->emit(getName(varDecl));
+    m_writer->emit(" : ");
+    emitType(elementType);
+    m_writer->emit(";\n");
+}
+
+void WGSLSourceEmitter::emitEntryPointAttributesImpl(
+    IRFunc* irFunc, IREntryPointDecoration* entryPointDecor
+    )
+{
+    auto stage = entryPointDecor->getProfile().getStage();
+
+    switch (stage)
+    {
+
+    case Stage::Fragment:
+        m_writer->emit("@fragment\n");
+        break;
+    case Stage::Vertex:
+        m_writer->emit("@vertex\n");
+        break;
+
+    case Stage::Compute:
+    {
+        m_writer->emit("@compute\n");
+
+        {
+            Int sizeAlongAxis[kThreadGroupAxisCount];
+            getComputeThreadGroupSize(irFunc, sizeAlongAxis);
+
+            m_writer->emit("@workgroup_size(");
+            for (int ii = 0; ii < kThreadGroupAxisCount; ++ii)
+            {
+                if (ii != 0)
+                    m_writer->emit(", ");
+                m_writer->emit(sizeAlongAxis[ii]);
+            }
+            m_writer->emit(")\n");
+        }
+    }
+    break;
+
+    default:
+        SLANG_ABORT_COMPILATION("unsupported stage.");
+    }
+
+}
+
+// This is 'function_header' from the WGSL specification
+void WGSLSourceEmitter::emitFuncHeaderImpl(IRFunc* func)
+{
+    Slang::IRType * resultType = func->getResultType();
+    auto name = getName(func);
+
+    m_writer->emit("fn ");
+    m_writer->emit(name);
+
+    emitSimpleFuncParamsImpl(func);
+
+    // An absence of return type is expressed by skipping the optional '->' part of the
+    // header.
+    if (resultType->getOp() != kIROp_VoidType)
+    {
+        m_writer->emit(" -> ");
+        emitType(resultType);
+    }
+}
+
+void WGSLSourceEmitter::emitSimpleFuncParamImpl(IRParam* param)
+{
+    if (auto sysSemanticDecor = param->findDecoration<IRTargetSystemValueDecoration>())
+    {
+        m_writer->emit("@builtin(");
+        m_writer->emit(sysSemanticDecor->getSemantic());
+        m_writer->emit(")");
+    }
+
+    CLikeSourceEmitter::emitSimpleFuncParamImpl(param);
+}
+
+void WGSLSourceEmitter::emitTransposedMatrixType(
+    IRType *const elementType, IRMatrixType *const matrixType
+    )
+{
+    IRIntegerValue const colCountSlang {getIntVal(matrixType->getColumnCount())};
+    IRIntegerValue const rowCountSlang {getIntVal(matrixType->getRowCount())};
+    IRIntegerValue const colCountWGSL {rowCountSlang};
+    IRIntegerValue const rowCountWGSL {colCountSlang};
+    emitMatrixType(elementType, rowCountWGSL, colCountWGSL);
+}
+
+void WGSLSourceEmitter::emitMatrixType(
+    IRType *const elementType, IRIntegerValue const& rowCountWGSL,
+    IRIntegerValue const& colCountWGSL
```

Comment:
For most part of the codebase, we are sticking to `const T*` or `const T&`, so let's stick with this convention if `const` is needed.

Reply:
Done. I tried to find and fix all other instances as well.